modules/programs/zsh/default.nix

```diff
@@ -0,0 +1,99 @@
+      inherit config;
+      name = "zsh";
+      includeTheme = true;
+      # TODO: copied from home-manager, fix this!
+      extraOptions = {
+        # TODO: Consider taking one from NixOS because of the strategy feature!
+        enableAutosuggestions = mkOption {
+          default = true;
+          type = types.bool;
+          description = "Enable zsh autosuggestions";
+        };
+        plugins = mkOption {
+          #
+          type = types.listOf soxin.lib.modules.zsh.pluginModule;
+          default = [ ];
+          example = literalExample ''
+            [
+              {
+                # will source zsh-autosuggestions.plugin.zsh
+                name = "zsh-autosuggestions";
+                src = pkgs.fetchFromGitHub {
+                  owner = "zsh-users";
+                  repo = "zsh-autosuggestions";
+                  rev = "v0.4.0";
+                  sha256 = "0z6i9wjjklb4lvr7zjhbphibsyx51psv50gm07mbb0kj9058j6kc";
+                };
+              }
+              {
+                name = "enhancd";
+                file = "init.sh";
+                src = pkgs.fetchFromGitHub {
+                  owner = "b4b4r07";
+                  repo = "enhancd";
+                  rev = "v2.2.1";
+                  sha256 = "0iqa9j09fwm6nj5rpip87x3hnvbbz9w9ajgm6wkrd5fls8fn8i5g";
+                };
+              }
+            ]
+          '';
+          description = "Plugins to source in <filename>.zshrc</filename>.";
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.enable (mkMerge [
+    # add all plugins installed by themes
+    { soxin.programs.zsh.plugins = cfg.theme.plugins; }
+
+    { programs.zsh = { inherit (cfg) enable; }; }
+
+    # Forward configurations to home-manager.
+    (optionalAttrs (mode == "home-manager") {
+      programs.zsh = { inherit (cfg) enableAutosuggestions plugins; };
+    })
+
+    # Forward configurations to NixOS.
+    (optionalAttrs (mode == "NixOS") {
+      programs.zsh.autosuggestions.enable = cfg.enableAutosuggestions;
+    })
+
+    # Forward plugins to NixOS.
+    # Copy the plugin management from home-manager
+    # TODO: Send it upstream to NixOS.
+    (optionalAttrs (mode == "NixOS") (mkIf (cfg.plugins != [ ]) {
+      # Many plugins require compinit to be called
+      # but allow the user to opt out.
+      programs.zsh.enableCompletion = mkDefault true;
+
+      environment.etc =
+        foldl' (a: b: a // b) { }
+          (map (plugin: { "${pluginsDir}/${plugin.name}".source = plugin.src; })
+            cfg.plugins);
+
+      programs.zsh.shellInit = concatStrings (map
+        (plugin: ''
+          path+="/etc/${pluginsDir}/${plugin.name}"
+          fpath+="/etc/${pluginsDir}/${plugin.name}"
+        '')
+        cfg.plugins);
+    }))
+
+    # install all completions libraries for system packages
+    (optionalAttrs (mode == "NixOS") (mkIf config.programs.zsh.enableCompletion {
+      environment.pathsToLink = [ "/share/zsh" ];
```

Comment:
This has caused me some loooong shell startup times, mainly because of `kubectl` completion. I haven't gotten around to fixing this, but it's something to keep in mind.

Reply:
Oh interesting, I didn't notice that. Perhaps requires some investigation from NixOS side.

Reply:
It's not blocking this PR though.